### PR TITLE
Take advantage of simplifications enabled by folding all indirect access to locals

### PR DESCRIPTION
**Pull request by SingleAccretion**
We delete code which needed to deal with the fact tracked locals in HIR could be accesses indirectly.

Diffs are primarily from the reintroduced morph folding. Sources:
1) Late morph in non-null assertion prop does not set `GTF_ORDER_SIDEEFF` on the local field (as it would on an indirection), allowing for some `optOptimizeBools` action.
2) CSE is less aggressive because `LCL_FLD` is cheaper than `IND(LCL_VAR_ADDR)`.
3) Args sorting is a bit different for the same reason (source of zero-sized diffs).

There is also a nice TP win, `0.3%` - `0.5%` reduction in instruction retired under FullOpts.

Note: diff better viewed with the "ignore whitespace changes" option.

### CODE REVIEW

**Comment by SingleAccretion** [comment]
@dotnet/jit-contrib

File: src/coreclr/jit/liveness.cpp [hunk 1]
```diff
@@ -240,18 +240,7 @@ void Compiler::fgPerNodeLocalVarLiveness(GenTree* tree)
...
-                if (!addrArg->DefinesLocalAddr(&lclVarTree))
-                {
-                    fgCurMemoryUse |= memoryKindSet(GcHeap, ByrefExposed);
-                }
-                else
-                {
-                    // Defines a local addr
-                    assert(lclVarTree != nullptr);
-                    fgMarkUseDef(lclVarTree->AsLclVarCommon());
-                }
```
Comment (jakobbotsch) [inline src/coreclr/jit/liveness.cpp:254]: :+1:

File: src/coreclr/jit/importer.cpp [hunk 2]
```diff
@@ -520,10 +520,9 @@ void Compiler::impAppendStmt(Statement* stmt, unsigned chkLevel, bool checkConsu
 
                 assert(retBuf->TypeIs(TYP_I_IMPL, TYP_BYREF));
 
-                GenTreeLclVarCommon* lclNode = retBuf->IsLocalAddrExpr();
-                if (lclNode != nullptr)
+                if (retBuf->OperIs(GT_LCL_VAR_ADDR))
```
Comment (jakobbotsch) [inline src/coreclr/jit/importer.cpp:523]: Can this not be `GT_LCL_FLD_ADDR`?

**Review by jakobbotsch — APPROVED** [review]
Looks great to me!

File: src/coreclr/jit/importer.cpp (hunk 2 above)
Comment (SingleAccretion) [reply src/coreclr/jit/importer.cpp:523]: No, only `LCL_VAR_ADDR`, currently at least.

It reflects the tight coupling this code has with what the importer actually produces for `stloc`. We don't look for indirect stores here for the same reason.

**Comment by jakobbotsch** [comment]
Failure is known according to build analysis.

**Pull request merged**